Add discarded discovery events ADR (#58)
trento-project/docs: adr/0016-discarded-discovery-events.md

# 16. Discarded Discovery Events

Date: 2024-10-15

## Status

Accepted

## Context

The discovery workflow allows Trento Agent instances to publish their host system's information to the Trento Web's `collect` API operation. Once accepted, the event results in one or more commands that are dispatched to the ES/CQRS system.

There are cases in which a discovery event sent from a Trento Agent is not processed; they can be grouped as:
1. The Trento Agent fails to connect to the Trento Web instance (network issues);
2. The Trento Agent fails to authenticate with the Trento Web instance (authentication issues);

> 3. The request payload is not in the expected shape, thus it cannot be mapped to a command (API contract issues);

Reviewer: ```suggestion
3. The request payload does not match the expected shape, thus it cannot be mapped to a command (API contract issues);
```

> 4. The discovery is rejected because of the current state in Trento.

Reviewer: ```suggestion
4. The discovery is rejected because of the current application state.
```

Reviewer: I tried to come up with a third choice but I just think `state in Trento` is ok then. Let's go!

# 16. Discarded Discovery Events

Date: 2024-10-15

## Status

Accepted

## Context

> The discovery workflow allows Trento Agent instances to publish their host system's information to the Trento Web's `collect` API operation. Once accepted, the event results in one or more commands that are dispatched to the ES/CQRS system.
> 
> There are cases in which a discovery event sent from a Trento Agent is not processed; they can be grouped as:

Reviewer 2: ```suggestion
There are cases in which a discovery sent from a Trento Agent is not processed; they can be categorized as:
```

Author: I like the rephrase. Do you want to remove the first sentence, too, or it's just a typo?

Reviewer 2: My suggestion is to omit that, because _it is known what a discovery is_.
However feel free to keep it, if it helps as an extra context. It does not harm anyway.

Author: I'd rather keep it to give a bit of context to occasional readers. It's just a sentence and it brings no much noise IMHO

1. The Trento Agent fails to connect to the Trento Web instance (network issues);
2. The Trento Agent fails to authenticate with the Trento Web instance (authentication issues);
3. The request payload is not in the expected shape, thus it cannot be mapped to a command (API contract issues);
4. The event is rejected because of the current data in Trento.

> Given the `collect` API is an HTTP endpoint, publishing Trento Agents receive a response according to the result of the operation.
> 
> The discovery workflow is eventually consistent, as the dispatch operation completes without waiting for the command to be processed and read models to be updated. 
> 
> As Trento Agents publish discovery events periodically, regardless of whether their data changes, the Trento system tends to converge and gather all relevant data, even if some events are not ingested.

Reviewer 2: ```suggestion
Currently all these scenarios result in an error being returned by the `collect` endpoint and subsequent log in trento agent.
```

## Decision

> Failures that depend on the current Trento data (4) are not reported as errors to the publishing Trento Agent; instead, it will receive a success status code indicating that the event has been correctly ingested by Trento Web, even if it won't be processed.

Reviewer 2: ```suggestion
Failures that depend on the current Trento data (4) are not reported as errors to the publishing Trento Agent; instead, it will receive a success status code indicating that the event has been correctly ingested by Trento Web, even if it won't be processed.
This is legitimated by the periodicity of the discoveries that would lead Trento to eventually converge to a consistent state.
```

Such events, alongside the ones that fall into (3), are traced into the `discarded_discovery_events` database table for inspection and troubleshooting.

The decision of whether API contract issues (3) should be treated the same way, has been postponed.

## Consequences

> * The separation of responsibilities between the publisher (Trento Agent) and the collector (Trento Web) is enforced: Trento Agent's job is done once the payload has been received and stored.

Reviewer 2: ```suggestion
```
There has been no actual separation enforcement.

> Trento Agent's job is done once the payload has been received
>
This was true also before the decision. The main difference is less failure logs in the agent, moved in trento web.

Author: > This was true also before the decision

I agree that the core principle is well established. Nevertheless, sometimes its implementation is subject to the developers' interpretation (like discovery error handling).  This ADR aims to clarify one more aspect on the core principle.

In this sense, iIused the term _enforce_. Should we say _clarify_? Also, we can move this sentence from consequences to context. Anyway I like the fact that the principle is stated once again.

Reviewer 2: Moving a revisited version of the statement to the context, highlighting that the agent does not really care knowing about discoveries failing because of server's internal state, is a good idea.

Author: Addressed in https://github.com/trento-project/docs/pull/58/commits/c10a0daca5cf79ba5d1401958f251a9bf5dca86a
